### PR TITLE
Lazy merkle tree merkle proof logic

### PR DESCRIPTION
**Pull request by vimwitch**
## Does this introduce a breaking change?

-   [ ] Yes
-   [ ] No

### CODE REVIEW

**Review by alrevuelta — APPROVED** [review]
lgtm! assuming tests pass locally.

File: packages/imt.sol/contracts/internal/InternalLazyIMT.sol
Comment (alrevuelta) [inline packages/imt.sol/contracts/internal/InternalLazyIMT.sol]: leftover comment?

File: packages/imt.sol/contracts/internal/InternalLazyIMT.sol
Comment (alrevuelta) [inline packages/imt.sol/contracts/internal/InternalLazyIMT.sol:226]: not an expert in solidity. any benefit modifying the value like this (pass by reference like) instead of returning it?

File: packages/imt.sol/contracts/internal/InternalLazyIMT.sol
Comment (chancehudson) [reply packages/imt.sol/contracts/internal/InternalLazyIMT.sol]: fixed

File: packages/imt.sol/contracts/internal/InternalLazyIMT.sol
Comment (chancehudson) [reply packages/imt.sol/contracts/internal/InternalLazyIMT.sol:226]: Yes, if we return an array from `_levels` then in this function we have to have 2 arrays. If we return the array we can't use it as `_elements` directly because we can't control the size and `_levels` might return an array that is smaller than we need, depending on the state of the tree.

It also removes a [copy operation/case](https://github.com/alrevuelta/zk-kit/pull/1/commits/2cd523cb53911277c727307a05ca8bbcaee65859#diff-656eaf8c59b6cf140813662fd12d464db2ada213c4246caefee1ea69bfa00eceR245) in the logic below. When I was returning the array from `_levels` the tests took longer on average than the original implementation, this change makes the tests faster. Test timing isn't a great metric for performance, but i _think_ the gas cost should also be lower because less memory is used with fewer copies.

**Comment by alrevuelta** [comment]
lgtm on my side. @vimwitch Ready to be merged? Unsure if you have permission since it's in my fork.

**Comment by vimwitch** [comment]
Yes, we'll merge this into your fork, then your fork into `zk-kit/main`

**Comment by alrevuelta** [comment]
@vimwitch Anything left here? lgtm on my side.

**Comment by vimwitch** [comment]
I think it's ready, this is your repo though so I can't merge it.

**Pull request merged**
